packages/jupyter-chat/src/input-model.ts

```diff
@@ -318,13 +318,46 @@ export class InputModel implements IInputModel {
    * Add attachment to send with next message.
    */
   addAttachment = (attachment: IAttachment): void => {
+    // Use JSON comparison to detect duplicates (same as in YChat.setAttachment)
+    const attachmentJson = JSON.stringify(attachment);
     const duplicateAttachment = this._attachments.find(
-      att => att.type === attachment.type && att.value === attachment.value
+      att => JSON.stringify(att) === attachmentJson
     );
     if (duplicateAttachment) {
       return;
     }
 
+    // Merge cells from same notebook into the same attachment
+    if (attachment.type === 'notebook' && attachment.cells) {
+      const existingNotebookIndex = this._attachments.findIndex(
+        att => att.type === 'notebook' && att.value === attachment.value
+      );
+
+      if (existingNotebookIndex !== -1) {
+        const existingAttachment = this._attachments[
+          existingNotebookIndex
+        ] as INotebookAttachment;
+        const existingCells = existingAttachment.cells || [];
+
+        // Filter out duplicate cells
+        const newCells = attachment.cells.filter(
+          newCell =>
+            !existingCells.some(existingCell => existingCell.id === newCell.id)
```

Comment:
this is n * m complexity, which may be okay if m (== existing cells) remain small.
If not, consider a mapping to a set here.